### PR TITLE
LinkLanguage refactor towards git-like syncing

### PR DESCRIPTION
**Pull request by lucksus**

### CODE REVIEW

File: src/language/Language.ts [hunk 1]
```diff
@@ -103,22 +104,36 @@ export interface GetAllAdapter {
     getAll(filter: any, count: number, page: number): Promise<Expression[] | null>;
 }
 
-export type NewLinksObserver = (added: Expression[], removed: Expression[])=>void;
+export type NewDiffObserver = (diff: PerspectiveDiff, revision: string)=>void;
```
Comment (kaichaosun) [inline src/language/Language.ts]: How about name it like `PerspectiveDiffObserver`?

File: src/language/Language.ts [hunk 2]
```diff
@@ -103,22 +104,36 @@ export interface GetAllAdapter {
...
+    // Aggregates all changes/diffs between the given revisions and
+    // render into one diff that the executor applies to the local
+    // perspective cache.
+    pullDiff(fromRev: string, toRev: string): Promise<PerspectiveDiff>;
 
-    // Get push notified by added links
-    addCallback(callback: NewLinksObserver);
+    // Publish changes.
+    // Returns new revision after adding diff.
+    commit(diff: PerspectiveDiff): Promise<string>;
```
Comment (kaichaosun) [inline src/language/Language.ts:135]: Commit seems local based operation, you probably mean push here?

**Review by kaichaosun — COMMENTED** [review]
I really like the idea, keen to know how is the latest revision gets calculated and validated in a decentralized way.

File: src/language/Language.ts (hunk 1 above)
Comment (lucksus) [reply src/language/Language.ts]: Yeah, makes more sense.

**Comment by lucksus** [comment]
@kaichaosun
> I really like the idea, keen to know how is the latest revision gets calculated and validated in a decentralized way.

@jdeepee and I are looking into using https://github.com/holochain-open-dev/holochain-time-index as time-based anchor to link new diffs off of.

File: src/links/Links.ts [hunk 3]
```diff
@@ -48,15 +48,3 @@ export function linkEqual(l1: LinkExpression, l2: LinkExpression): boolean {
...
-    const mash = JSON.stringify(link.data, Object.keys(link.data).sort()) +
-                JSON.stringify(link.author) + link.timestamp
-    let hash = 0, i, chr;
-    for (i = 0; i < mash.length; i++) {
-        chr   = mash.charCodeAt(i);
-        hash  = ((hash << 5) - hash) + chr;
-        hash |= 0; // Convert to 32bit integer
-    }
-    return hash;
-}
```
Comment (lucksus) [inline src/links/Links.ts:62]: Why remove these lines?

**Pull request merged**